cpp/src/interpolate/cubic_spline.cu

```diff
@@ -0,0 +1,404 @@
+        thrust::make_counting_iterator<int>(1),
+        thrust::make_counting_iterator<int>(prefixes.size()),
+        [T_, Y_, PREFIXES, H_, I_, Z_, D3_, D2_, D1_, D0_] __device__
+        (int index) {
+          int n = PREFIXES[index] - PREFIXES[index-1];
+          int h = PREFIXES[index-1];
+          int dh = PREFIXES[index-1] - (index-1);
+          int ci = 0;
+          for(ci = 0 ; ci < n-1 ; ++ci) {
+            T a = Y_[h+ci];
+            T b = I_[h+ci] - H_[h+ci] * (Z_[h+ci+1] + 2 * Z_[h+ci]) / 6;
+            T c = Z_[h+ci] / 2.0;
+            T d = (Z_[h+ci+1] - Z_[h+ci]) / 6 * H_[h+ci];
+            T t = T_[h+ci];
+            D3_[dh+ci] = d;
+            D2_[dh+ci] = c - 3 * d * t;
+            D1_[dh+ci] = b - t * (2*c - t * (3 * d));
+            D0_[dh+ci] = a - t * (b - t * (c - t * d)); // horners
+          }
+      });
+  };
+  template<typename T>
+  std::enable_if_t<not std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& y,
+                  cudf::column_view const& prefixes,
+                  cudf::mutable_column_view const& h,
+                  cudf::mutable_column_view const& i,
+                  cudf::mutable_column_view const& z,
+                  cudf::mutable_column_view const& d3,
+                  cudf::mutable_column_view const& d2,
+                  cudf::mutable_column_view const& d1,
+                  cudf::mutable_column_view const& d0,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      CUDF_FAIL("Non-floating point operation is not supported.");
+  }
+};
+
+struct compute_spline_tridiagonals {
+  template<typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& y,
+                  cudf::column_view const& prefixes,
+                  cudf::mutable_column_view const& D,
+                  cudf::mutable_column_view const& Dlu,
+                  cudf::mutable_column_view const& u,
+                  cudf::mutable_column_view const& h,
+                  cudf::mutable_column_view const& i,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      const T* T_ = t.data<T>();
+      const T* Y_ = y.data<T>();
+      const int32_t* PREFIXES = prefixes.data<int32_t>();
+      T* D_ = D.data<T>();
+      T* Dlu_ = Dlu.data<T>();
+      T* U_ = u.data<T>();
+      T* H_ = h.data<T>();
+      T* I_ = i.data<T>();
+      thrust::for_each(rmm::exec_policy(stream)->on(stream),
+        thrust::make_counting_iterator<int>(1),
+        thrust::make_counting_iterator<int>(prefixes.size()),
+        [T_, Y_, PREFIXES, D_, Dlu_, U_, H_, I_] __device__
+        (int index) {
+          int n = PREFIXES[index] - PREFIXES[index-1];
+          int h = PREFIXES[index-1];
+          int ci = 0;
+          for(ci = 0 ; ci < n-1 ; ++ci) {
```

Comment:
Why can't you parallelize the for_each over the total number of elements and eliminate these inner loops -- go fully parallel.  All the inner loop iterations are independent, save for this statement: 

`I_[h + ci] = (Y_[h+ci+1] - Y_[h+ci]) / H_[h + ci];`

And that can be parallelized because it depends on the value of `H_[h+ci]` which is computed in the same iteration -- just store it in a variable for clarity.